Mathlib/RingTheory/TensorProduct/IsBaseChangeFree.lean

```diff
@@ -64,10 +65,82 @@ theorem basis_repr_comp (v : V) :
   ext i
   simp [basis_repr_comp_apply]
 
-variable [Module.Free R V]
-
 include ibc in
-theorem free : Module.Free S W :=
+theorem free [Module.Free R V] : Module.Free S W :=
   Module.Free.of_basis (ibc.basis (Module.Free.chooseBasis R V))
 
 end IsBaseChange
+
+section underring
+
+namespace IsBaseChange
+
+open TensorProduct
+
+variable {R : Type*} [CommSemiring R]
+  {V : Type*} [AddCommMonoid V] [Module R V]
+  (A : Type*) [CommSemiring A] [Algebra A R]
+  [Module A V] [IsScalarTower A R V]
+
+open TensorProduct
+
+variable {ι : Type*} (b : Module.Basis ι R V)
+
+theorem of_basis : IsBaseChange R (Finsupp.linearCombination A b) := by
+  have : DecidableEq ι := Classical.typeDecidableEq ι
+  let j := TensorProduct.finsuppScalarRight' A R ι R
+  refine of_equiv ?_ ?_
+  · apply LinearEquiv.ofBijective (Finsupp.linearCombination R b ∘ₗ j)
+    rw [LinearMap.coe_comp, LinearEquiv.coe_toLinearMap, j.bijective.of_comp_iff]
+    simp [Function.Bijective,
+        ← span_range_eq_top_iff_surjective_finsuppLinearCombination,
+        ← linearIndependent_iff_injective_finsuppLinearCombination,
+        Module.Basis.span_eq, b.linearIndependent]
+  · intro x
+    simp only [LinearEquiv.ofBijective_apply, LinearMap.coe_comp, LinearEquiv.coe_coe,
+      Function.comp_apply]
+    suffices (j (1 ⊗ₜ[A] x)) = x.mapRange (algebraMap A R) (by simp) by
+      rw [this]
+      simp only [Finsupp.linearCombination_apply]
+      rw [Finsupp.sum_mapRange_index (by simp)]
+      aesop
+    ext i
+    simp [j, TensorProduct.finsuppScalarRight', Algebra.algebraMap_eq_smul_one]
+
+include A in
+/-- Any finite basis of a module can express it as the base change
+of a finite free module from any under-ring. -/
+theorem of_fintype_basis [Fintype ι] :
+    IsBaseChange R (Fintype.linearCombination A b) := by
+  have : DecidableEq ι := Classical.typeDecidableEq ι
+  let j : R ⊗[A] (ι → A) ≃ₗ[R] ι → R := piScalarRight A R R ι
+  refine of_equiv ?_ ?_
+  · apply LinearEquiv.ofBijective (Fintype.linearCombination R b ∘ₗ j)
+    rw [LinearMap.coe_comp, LinearEquiv.coe_toLinearMap, j.bijective.of_comp_iff]
+    simp [Function.Bijective,
+        ← span_range_eq_top_iff_surjective_fintypeLinearCombination,
+        ← linearIndependent_iff_injective_fintypeLinearCombination,
+        Module.Basis.span_eq, b.linearIndependent]
+  · intro x
+    rw [LinearEquiv.ofBijective_apply, LinearMap.coe_comp, LinearEquiv.coe_coe,
+      Function.comp_apply, Fintype.linearCombination_apply,
+      Fintype.linearCombination_apply]
```

Comment:
```suggestion
    -- simp? [Fintype.linearCombination_apply] says:
    simp only [LinearEquiv.ofBijective_apply, LinearMap.coe_comp, LinearEquiv.coe_coe,
      Function.comp_apply, Fintype.linearCombination_apply]
```

This is up to you, and some people disagree with me, but I prefer (mostly) auto-generated simp only blocks to rw blocks, because it makes the proof more maintainable (if the proof breaks in the future, the `simp?` statement that generated it can sometimes help fix it (which is why I put it in a comment above).  